### ISSUE TITLE
[C++] compilation standard to C++20

### ISSUE THREAD

**Issue by rkannan** [apache/arrow#44792]
### Describe the enhancement requested

are there any issues with setting the default to c++20 in the codebase? Currently the cmake_modules/SetupCxxFlags.cmake sets CMAKE_CXX_STANDARD to 17. 

### Component(s)

C++

**Comment by assignUser** [comment]
Yes that's not a quick decision to make. Here is the thread about bumping to 17 2 years ago: https://lists.apache.org/thread/9g14n3odhj6kzsgjxr6k6d3q73hg2njr

**Comment by assignUser** [comment]
I checked and we currently can't switch to 20 because CRAN (R package hub) has several builds with AppleClang <=14. This is a blocker but not necessarily the only one.

You are certainly not alone in preferring to use C++20 but as arrow C++ is a foundational library we have to be more conservative with such thinks.

I will close this issue for now, we can re-visted in time

**Comment by kou** [comment]
It seems that "Apple Clang 1300.0.29.30" is the oldest version based on the last R arrow package check information:

* https://cran.r-project.org/web/checks/check_results_arrow.html
* https://cran.r-project.org/web/checks/check_flavors.html#r-oldrel-macos-x86_64

It's based on LLVM 12.0.0. See also: https://en.wikipedia.org/wiki/Xcode#Xcode_11.0_-_14.x_(since_SwiftUI_framework)_2

LLVM 12.0.0 has `-std=c++20`. See also: https://clang.llvm.org/cxx_status.html#cxx20

So we may be able to use C++20 on CRAM. (We may not be able to use some C++20 features.)

**Comment by kou** [comment]
We can use C++20 with GCC 8 or later: https://gcc.gnu.org/projects/cxx-status.html#cxx20

**Comment by assignUser** [comment]
I think before such a move we should properly document our support policy. That will create clarity both for users and contributors.

**Comment by kou** [comment]
> I think before such a move we should properly document our support policy.

It mentions supported GCC/Clang/MSVC versions explicitly, right?

**Comment by assignUser** [comment]
AFAIK this is the only place we mention it:
>A C++17-enabled compiler. On Linux, gcc 7.1 and higher should be sufficient. For Windows, at least Visual Studio VS2017 is required.

https://arrow.apache.org/docs/developers/cpp/building.html

**Comment by kou** [comment]
It seems that we need to add only Clang version to the document.

**Comment by kou** [comment]
FYI: We tried C++20: #45445

**Comment by pitrou** [comment]
Thanks for pointing me to this issue @kou

**Comment by rok** [comment]
We now compile to C++20, see https://github.com/apache/arrow/issues/48587
